### PR TITLE
[LoRA] Remove legacy LoRA code and related adjustments

### PR DESCRIPTION
**Pull request by sayakpaul**
# What does this PR do?

It's been a while since we have had legacy LoRA code, and we have no reason to continue to include them within the library, given that `peft` is a required dependency for all things LoRA within the library. 

This PR removes all the legacy code from `loaders/lora.py` and `loaders/unet.py` and revises the `load_attn_procs()` method in the UNet. 

Additionally, I have run all the slow LoRA tests to ensure they're passing. 

In summary, this PR should make our LoRA-related code blocks much cleaner and easier to navigate through.

### CODE REVIEW

**Comment by sayakpaul** [comment]
Requesting for reviews from @younesbelkada and @BenjaminBossan initially. Once done, will ping others from the team.

File: .github/workflows/pr_test_peft_backend.yml [hunk 1]
```diff
@@ -111,3 +111,21 @@ jobs:
...
+      run: |
+        cat reports/tests_${{ matrix.config.report }}_failures_short.txt
+        cat reports/tests_${{ matrix.config.report }}_models_lora_failures_short.txt
+    
+    - name: Test suite reports artifacts
+      if: ${{ always() }}
+      uses: actions/upload-artifact@v2
+      with:
+        name: pr_${{ matrix.config.report }}_test_reports
+        path: reports
```
Comment (sayakpaul) [inline .github/workflows/pr_test_peft_backend.yml:131]: We should upload the test reports for LoRA.

File: src/diffusers/loaders/lora.py [hunk 2]
```diff
@@ -625,9 +519,7 @@ def load_lora_into_text_encoder(
                 # Unsafe code />
 
     @classmethod
-    def load_lora_into_transformer(
-        cls, state_dict, network_alphas, transformer, low_cpu_mem_usage=None, adapter_name=None, _pipeline=None
```
Comment (sayakpaul) [inline src/diffusers/loaders/lora.py:629]: We have absolutely no reason to continue to include `low_cpu_mem_usage` when using `peft`.

File: src/diffusers/loaders/lora.py [hunk 3]
```diff
@@ -395,24 +378,18 @@ def load_lora_into_unet(
...
             adapter_name (`str`, *optional*):
                 Adapter name to be used for referencing the loaded adapter model. If not specified, it will use
                 `default_{i}` where i is the total number of adapters being loaded.
         """
         if not USE_PEFT_BACKEND:
             raise ValueError("PEFT backend is required for this method.")
 
-        from peft import LoraConfig, inject_adapter_in_model, set_peft_model_state_dict
-
-        low_cpu_mem_usage = low_cpu_mem_usage if low_cpu_mem_usage is not None else _LOW_CPU_MEM_USAGE_DEFAULT
```
Comment (sayakpaul) [inline src/diffusers/loaders/lora.py:415]: `low_cpu_mem_usage` isn't used when using `peft`.

File: src/diffusers/loaders/lora.py [hunk 4]
```diff
@@ -846,22 +731,11 @@ def unload_lora_weights(self):
...
-            if version.parse(__version__) > version.parse("0.23"):
-                logger.warning(
-                    "You are using `unload_lora_weights` to disable and unload lora weights. If you want to iteratively enable and disable adapter weights,"
-                    "you can use `pipe.enable_lora()` or `pipe.disable_lora()`. After installing the latest version of PEFT."
-                )
+            raise ValueError("PEFT backend is required for this method.")
 
-            for _, module in unet.named_modules():
-                if hasattr(module, "set_lora_layer"):
-                    module.set_lora_layer(None)
```
Comment (sayakpaul) [inline src/diffusers/loaders/lora.py:860]: Legacy code. Shouldn't stay.

File: src/diffusers/loaders/unet.py [hunk 5]
```diff
@@ -216,198 +206,187 @@ def load_attn_procs(self, pretrained_model_name_or_path_or_dict: Union[str, Dict
...
-
-        is_lora = all(("lora" in k or k.endswith(".alpha")) for k in state_dict.keys()) and not USE_PEFT_BACKEND
         is_custom_diffusion = any("custom_diffusion" in k for k in state_dict.keys())
+        is_lora = all(("lora" in k or k.endswith(".alpha")) for k in state_dict.keys())
+        is_model_cpu_offload = False
+        is_sequential_cpu_offload = False
 
-        if is_lora:
-            # correct keys
-            state_dict, network_alphas = self.convert_state_dict_legacy_attn_format(state_dict, network_alphas)
```
Comment (sayakpaul) [inline src/diffusers/loaders/unet.py:227]: No need as already handled in `peft` and the backend.

File: src/diffusers/loaders/unet.py [hunk 6]
```diff
@@ -216,198 +206,187 @@ def load_attn_procs(self, pretrained_model_name_or_path_or_dict: Union[str, Dict
...
+        if is_custom_diffusion:
+            attn_processors = self._process_custom_diffusion(state_dict=state_dict)
+        elif is_lora:
+            is_model_cpu_offload, is_sequential_cpu_offload = self._process_lora(
+                state_dict=state_dict,
+                unet_identifier_key=self.unet_name,
+                network_alphas=network_alphas,
+                adapter_name=adapter_name,
+                _pipeline=_pipeline,
+            )
```
Comment (sayakpaul) [inline src/diffusers/loaders/unet.py:223]: `_process_lora()` injects the LoRA state dict inline.

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks a lot for taking care of removing the legacy LoRA code. Always nice to see more code being removed than added: +388 −425

Overall, this LGTM, from my understanding, most changes are about removing the legacy code paths and moving some LoRA-related code around without actually changing the logic. The tests all passing is the best indicator that nothing was broken.

To me, it is unclear if the removal of the `low_cpu_mem_usage` and `mirror` options are in any way related to the refactor. If not, maybe in the future it would be better to split these types of PRs to make them easier to review.

File: src/diffusers/loaders/unet.py [hunk 7]
```diff
@@ -216,198 +206,189 @@ def load_attn_procs(self, pretrained_model_name_or_path_or_dict: Union[str, Dict
...
 
-        # <Unsafe code
-        # We can be sure that the following works as it just sets attention processors, lora layers and puts all in the same dtype
-        # Now we remove any existing hooks to
-        is_model_cpu_offload = False
-        is_sequential_cpu_offload = False
+        return attn_processors
 
-        # For PEFT backend the Unet is already offloaded at this stage as it is handled inside `load_lora_weights_into_unet`
+    def _process_lora(self, state_dict, unet_identifier_key, network_alphas, adapter_name, _pipeline):
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply src/diffusers/loaders/unet.py:282]: See if it makes sense now.

**Comment by sayakpaul** [comment]
@BenjaminBossan thanks for the comments. I have addressed all of them or have replied to them. 

> To me, it is unclear if the removal of the low_cpu_mem_usage and mirror options are in any way related to the refactor. If not, maybe in the future it would be better to split these types of PRs to make them easier to review.

I fully agree with you here. I don't like increasing the scope of PR review, either. However, I decided to include the changes related to `low_cpu_mem_usage` because it's actually not used in the peft part of the code. So, I thought it's related. 

LMK if it makes sense.

**Review by BenjaminBossan — APPROVED** [review]
Sorry, I forgot to reply to your reply. Thanks for making the adjustments and again big thanks for taking care of the legacy code :rocket:

**Comment by sayakpaul** [comment]
@yiyixuxu ready for a review now that I have an approval from Benjamin.

**Review by yiyixuxu — APPROVED** [review]
ok to merge because all the tests passed and @BenjaminBossan approved!
